backend/src/main/java/fk/prof/backend/aggregator/AggregationWindow.java

```diff
@@ -0,0 +1,159 @@
+    if (workInfo.getStatus().equals(AggregationStatus.SCHEDULED)) {
+      workInfo.setStatus(AggregationStatus.ONGOING);
+    } else if (workInfo.getStatus().equals(AggregationStatus.PARTIAL)) {
+      workInfo.setStatus(AggregationStatus.ONGOING_PARTIAL);
+    } else {
+      updated = false;
+    }
+
+    if (updated) {
+      workInfo.setStartedAt(startedAt);
+    }
+
+    return updated;
+  }
+
+  public boolean completeProfile(long workId) {
+    boolean updated = true;
+    ProfileWorkInfo workInfo = this.workStatusLookup.get(workId);
+    if (workInfo.getStatus().equals(AggregationStatus.ONGOING)) {
+      workInfo.setStatus(AggregationStatus.COMPLETED);
+    } else if (workInfo.getStatus().equals(AggregationStatus.ONGOING_PARTIAL)) {
+      workInfo.setStatus(AggregationStatus.RETRIED);
+    } else {
+      updated = false;
+    }
+
+    if (updated) {
+      workInfo.setEndedAt(LocalDateTime.now(Clock.systemUTC()));
+    }
+    return updated;
+  }
+
+  public boolean abandonProfile(long workId) {
+    boolean updated = true;
+    ProfileWorkInfo workInfo = this.workStatusLookup.get(workId);
+    if (workInfo.getStatus().equals(AggregationStatus.ONGOING) || workInfo.getStatus().equals(AggregationStatus.ONGOING_PARTIAL)) {
+      workInfo.setStatus(AggregationStatus.PARTIAL);
+    } else {
+      updated = false;
+    }
+
+    if (updated) {
+      workInfo.setEndedAt(LocalDateTime.now(Clock.systemUTC()));
+    }
+    return updated;
+  }
+
+  /**
+   * Aborts all in-flight profiles. Should be called when aggregation window expires
+   */
+  public void abortOngoingProfiles() {
+    this.workStatusLookup.replaceAll((workId, workInfo) -> {
+      if (workInfo.getStatus().equals(AggregationStatus.ONGOING) || workInfo.getStatus().equals(AggregationStatus.ONGOING_PARTIAL)) {
+        workInfo.setStatus(AggregationStatus.ABORTED);
+      }
+      return workInfo;
+    });
+  }
+
+  public void aggregate(Recorder.Wse wse, RecordedProfileIndexes indexes) throws AggregationFailure {
+    switch (wse.getWType()) {
+      case cpu_sample_work:
+        Recorder.StackSampleWse stackSampleWse = wse.getCpuSampleEntry();
+        if (stackSampleWse == null) {
+          throw new AggregationFailure(String.format("work type=%s did not have associated samples", wse.getWType()));
+        }
+        cpuSamplingAggregationBucket.aggregate(stackSampleWse, indexes);
+        break;
+      default:
+        throw new AggregationFailure(String.format("Aggregation not supported for work type=%s", wse.getWType()));
+    }
+  }
+
+  public void updateWorkInfo(long workId, Recorder.Wse wse) {
+    ProfileWorkInfo workInfo = getWorkInfo(workId);
+    if (workInfo == null) {
+      throw new AggregationFailure(String.format("Cannot find work id=%d association in the aggregation window", workId), true);
+    }
+
+    for (Recorder.TraceContext trace : wse.getIndexedData().getTraceCtxList()) {
+      workInfo.addTrace(trace.getTraceName(), trace.getCoveragePct());
```

Comment:
Update: We had an offline discussion and the scenario that came up was this: Recorder can timeout on /profile API request and make a new request for same work id to /profile. Or there can be a rogue recorder/some other bug which can lead to different profile API requests for same work-id. To prevent such scenarios, we have added a check which ensures that at any given time, if multiple /profile requests are received by backend for same work-id, only the first one will be processed, and rest will be discarded.